device/transport/mqtt/src/mqtt_twin_client.ts

```diff
@@ -0,0 +1,266 @@
+                this.transition(client, 'subscribed', callback);
+              }
+            });
+          },
+          '*': function (client: TopicSubscription): void { this.deferUntilTransition(client); }
+        },
+        subscribed: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            callback();
+          },
+          subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            callback();
+          },
+          unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            this.transition(client, 'unsubscribing', callback);
+          }
+        },
+        unsubscribing: {
+          _onEnter: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+            client.mqttClient.unsubscribe(client.topic, (err, result) => {
+              if (err) {
+                debug('failed to unsubscribe: ' + err.toString());
+              } else {
+                debug('unsubscribed from: ' + client.topic);
+              }
+              this.transition(client, 'unsubscribed', err, callback);
+            });
+          },
+          '*': function (client: TopicSubscription): void { this.deferUntilTransition(client); }
+        }
+      },
+      subscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+        this.handle(client, 'subscribe', callback);
+      },
+      unsubscribe: function (client: TopicSubscription, callback: (err?: Error) => void): void {
+        this.handle(client, 'unsubscribe', callback);
+      }
+    });
+
+    this._responseTopic = {
+      mqttClient: this._mqtt,
+      topic: responseTopic
+    };
+
+    this._desiredPropertiesUpdatesTopic = {
+      mqttClient: this._mqtt,
+      topic: desiredPropertiesUpdatesTopic
+    };
+  }
+
+  getTwin(callback: (err?: Error, twin?: TwinProperties) => void): void {
+    /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_002: [The `getTwin` method shall subscribe to the `$iothub/twin/res/#` topic if not already subscribed.]*/
+    this._topicFsm.subscribe(this._responseTopic, (err) => {
+      if (err) {
+        /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_009: [If subscribing to the response topic fails, the callback shall be called with the translated version of the error obtained by using the `translateError` method of the `azure-iot-mqtt-base` package.]*/
+        callback(translateError(err));
+      } else {
+        /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_003: [The `getTwin` method shall publish the request message on the `$iothub/twin/get/?rid=<requestId>` topic using the `MqttBase.publish` method.]*/
+        /*Codes_SRS_NODE_DEVICE_MQTT_TWIN_CLIENT_16_006: [The request message published by the `getTwin` method shall have an empty body.]*/
+        this._sendTwinRequest('GET', '/', ' ', callback);
```

Comment:
>('GET', '/', ' ', callback) [](start = 29, length = 27)

A while back I had a pull request here to deal with the '/' parameter.  It would be nice to fix this now (maybe).  (It would have removed the use of '/' here.  Perhaps we should discuss this?

Reply:
yep - one of the points of this refactoring is to address this issue. since the resource paths are now stored in the transport there is no need to worry about that anymore. mqtt will do its thing and amqp will do its thing, the device client won't have to care about that anymore.

---
In reply to: [170687317](https://github.com/Azure/azure-iot-sdk-node/pull/236#discussion_r170687317) [](ancestors = 170687317)